### PR TITLE
More conversions

### LINKED ISSUE #956: Ability to short circuit ienum conversions to be able to handle b64 => byte[]
## Summary

Brief explanation of the feature.

I would like to have a simple way to have a conversion of type string => byte[]

currently this will decompose to an attempt to convert char => byte.

### Motivation

Why are we doing this? What use cases does it support? What is the expected outcome?

Allow for passing binary information as arguments

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #956

### CODE REVIEW

File: src/GraphQL/ObjectExtensions.cs [hunk 1]
```diff
@@ -103,6 +103,17 @@ public static object ToObject(this IDictionary<string, object> source, Type type
         /// <remarks>There is special handling for strings, IEnumerable&lt;T&gt;, Nullable&lt;T&gt;, and Enum.</remarks>
         public static object GetPropertyValue(this object propertyValue, Type fieldType, IGraphType mappedType = null)
         {
+            bool SkipEnumerable()
+            {
+                if (fieldType == typeof(string))
+                    return true;
+
+                if (fieldType == typeof(byte[]) && propertyValue is string)
```
Comment (sungam3r) [inline src/GraphQL/ObjectExtensions.cs]: I think this can be safely considered a special case.

File: src/GraphQL/Conversion/ValueConverter.cs [hunk 2]
```diff
@@ -114,6 +114,11 @@ static ValueConverter()
             Register(typeof(double), typeof(decimal), value => Convert.ToDecimal(value, NumberFormatInfo.InvariantInfo));
 
             Register(typeof(string), typeof(Uri), value => value is string s ? new Uri(s) : (Uri)value);
+            // registering such a default conversion for string->byte[] seems useful
+            Register(typeof(string), typeof(byte[]), value => Convert.FromBase64String((string)value));
```
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs]: @Shane32 This is worth noting in the documentation.

File: src/GraphQL/ObjectExtensions.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/ObjectExtensions.cs]: Good idea.

Comment (sungam3r) [reply src/GraphQL/ObjectExtensions.cs]: If we go further, we can make the assertion that `IEnumerable` has nothing to do with it, we can simply move `ValueConverter.ConvertTo(value, fieldType);`/lookup to the beginning of the method.

Comment (sungam3r) [reply src/GraphQL/ObjectExtensions.cs]: Done.

File: src/GraphQL/Conversion/ValueConverter.cs [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 using System.Globalization;
 using System.Numerics;
 
 namespace GraphQL
 {
     public static class ValueConverter
     {
-        private static readonly Dictionary<Type, Dictionary<Type, Func<object, object>>> _valueConversions
-            = new Dictionary<Type, Dictionary<Type, Func<object, object>>>();
+        private static readonly ConcurrentDictionary<Type, ConcurrentDictionary<Type, Func<object, object>>> _valueConversions
```
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs]: Accidentally noticed that this static field was still not thread safe. All other static dictionaries are `ConcurrentDictionary<>`.

File: src/GraphQL/Conversion/ValueConverter.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQL/Conversion/ValueConverter.cs]: Probably should be noted in `schema-types.md` as well as in the upgrade notes; would you agree?

Comment (sungam3r) [reply src/GraphQL/Conversion/ValueConverter.cs]: yes

Comment (sungam3r) [reply src/GraphQL/Conversion/ValueConverter.cs]: There will be another change with value conversion soon.

**Comment by Shane32** [comment]
I didn't read all the code in depth, but can you help me with a very quick overview of this here and be sure I understand this correctly?  It will help me write better documentation.

- When a input variable comes in via the variables, it gets converted by the proper ScalarGraphType to a .Net type, and then gets converted by ValueConverter to the type requested by GetArgument

- When an parameter comes in via the query (as a literal), it gets converted by the proper ScalarGraphType to a .Net type, and then gets converted by ValueConverter to the type requested by GetArgument

- When serializing response data, the .Net type gets passed directly to the proper ScalarGraphType, ignoring the ValueConverter, which handles the serialization?

- Or, when serializing response data, the ValueConverter always makes a first pass, before feeding the ScalarGraphType?  If this is true how does the ValueConverter need to know what graph type to prep for?

**Comment by Shane32** [comment]
And
- When GetArgument does not request a particular type, it skips the ValueConverter (I assume)

**Comment by Shane32** [comment]
If you can give me some quick pointers, it will probably save me a while of digging through the code to figure out how it all fits together - thanks @sungam3r

**Comment by sungam3r** [comment]
There will be no quick answer. I made these changes mostly for `GetArgument<T>`. I understand that `ValueConverter` is used not only for this, but has not yet verified all usages. I will come back with answers later.

**Comment by Shane32** [comment]
Ok thanks; I don't have time to dig into it now.  If you post what you learn, it will help when I do have time.  Thanks!

**Comment by sungam3r** [comment]
Relates to #1616

**Comment by sungam3r** [comment]
@Shane32 Regarding your questions.

1. When a input variable comes in via the variables, it gets converted by the proper `ScalarGraphType` to a .Net type, and then gets converted by `ValueConverter` to the type requested by `GetArgument`.

Yes.

2. When an parameter comes in via the query (as a literal), it gets converted by the proper `ScalarGraphType` to a .Net type, and then gets converted by `ValueConverter` to the type requested by `GetArgument`.

Yes.

3. When serializing response data, the .Net type gets passed directly to the proper `ScalarGraphType`, ignoring the `ValueConverter`, which handles the serialization?

No. Most `Serialize` methods delegate call to `ParseValue` which calls `ValueConverter.ConvertTo`. See https://github.com/graphql-dotnet/graphql-dotnet/blob/master/docs2/site/docs/getting-started/custom-scalars.md

4. Or, when serializing response data, the `ValueConverter` always makes a first pass, before feeding the `ScalarGraphType`? If this is true, does the `ValueConverter` need to know what graph type to prep for?

No. `Serialize` is used only in `ValueExecutionNode.ToValue()`:
```c#
  public override object ToValue()
        {
            if (Result == null)
                return null;

            var scalarType = GraphType as ScalarGraphType;
            return scalarType?.Serialize(Result);
        }
```
No. `ValueConverter` knows nothing about GraphType, it works with registered .NET -> .NET conversions.

5. When `GetArgument` does not request a particular type, it skips the `ValueConverter` (I assume).

Yes.
```c#
 if (argumentType == typeof(object))
                {
                    result = arg;
                    return true;
                }
```

**Pull request merged**